### ISSUE TITLE
Support for AWS Signature Version 4

### ISSUE THREAD

**Issue by michaelthorne** [adrianfalleiro/gulp-s3-publish#3]
This is also an issue in the [gulp-s3](https://github.com/nitaigao/gulp-s3/issues/62) plugin, but mentioning it here too.

I've been unable to upload to certain (newer) S3 regions and it is as a result of them _only_ supporting V4. i.e. `eu-west-2` (London) as a region, returns an ambiguous HTTP 400 error when attempting to publish files to a bucket in that region.

```
Message:
    [400] C:\file.css.gz -> file.css
Stack:
Error: [400] C:\file.css.gz -> file.css
    at DestroyableTransform.uploadCallback (C:\node_modules\gulp-s3-publish\index.js:76:24)
```

If I _only_ switch the region to e.g. `us-east-1` (formerly US Standard which supports V2 _and_ V4) and leave everything else as is, the publish works as expected.

**Comment by adrianfalleiro** [comment]
Cheers @michaelthorne, it looks like `knox` does not support the V4 Signature, and there's an un-merged PR thats been open the last 2 years to add support (https://github.com/Automattic/knox/pull/273).

Might need to refactor using `aws-sdk` or another streaming library like `knox`

**Comment by michaelthorne** [comment]
@adrianfalleiro that would be awesome!

**Comment by adrianfalleiro** [comment]
@michaelthorne can you try with the `master` branch and let me know if that works for you?

You can install it by running `npm install git+https://github.com/adrianfalleiro/gulp-s3-publish`

**Comment by michaelthorne** [comment]
@adrianfalleiro thanks for the quick update! I got the following error after installing it as per your comment above:

```
Message:
    TypeError: Cannot read property 'statusCode' of null
Stack:
Error: TypeError: Cannot read property 'statusCode' of null
    at DestroyableTransform.<anonymous> (C:\test\node_modules\gulp-s3-publish\index.js:89:32)
    at DestroyableTransform._transform (C:\test\node_modules\through2-concurrent\through2-concurrent.js:30:15)
    at DestroyableTransform.Transform._read (C:\test\node_modules\readable-stream\lib\_stream_transform.js:159:10)
    at DestroyableTransform.Transform._write (C:\test\node_modules\readable-stream\lib\_stream_transform.js:147:83)
    at doWrite (C:\test\node_modules\readable-stream\lib\_stream_writable.js:345:64)
    at writeOrBuffer (C:\test\node_modules\readable-stream\lib\_stream_writable.js:334:5)
    at DestroyableTransform.Writable.write (C:\test\node_modules\readable-stream\lib\_stream_writable.js:271:11)
    at write (C:\test\node_modules\gulp-gzip\node_modules\readable-stream\lib\_stream_readable.js:623:24)
    at flow (C:\test\node_modules\gulp-gzip\node_modules\readable-stream\lib\_stream_readable.js:632:7)
    at Transform.pipeOnReadable (C:\test\node_modules\gulp-gzip\node_modules\readable-stream\lib\_stream_readable.js:664:5)
```

FYI, using `eu-west-2` as a region.

**Comment by adrianfalleiro** [comment]
Cheers @michaelthorne. I've fixed that issue with the error messaging. However I haven't done any testing for uploading gzipped files so that might also be an issue.

I see gulp-gzip in that stack trace. Are you gzipping files?

**Comment by adrianfalleiro** [comment]
Ok got it working with gzip. Can you try again and let me know if working.

Cheers

**Comment by adrianfalleiro** [comment]
Tested and released on npm as 2.0.0.
